framework/data/ActiveDataProvider.php

```diff
@@ -97,23 +97,36 @@ public function init()
      * {@inheritdoc}
      */
     protected function prepareModels()
+    {
+        $query = $this->prepareQuery();
+
+        return $query ? $query->all($this->db) : [];
+    }
+
+    /**
+     * Prepares the sql-query that will get the data for current page.
+     * @return array|QueryInterface
```

Comment:
It probably should be `QueryInterface|null`. Or even just `QueryInterface` if you use `emulateExecution()` instead of returning `null`/`false`.

Reply:
I like this idea, but I think we also should add method `getEmulateExecution` for testing purposes.

Reply:
Why?

Reply:
https://github.com/yiisoft/yii2/blob/5544d31bb3b7b0959ef51f7c78c7e06c8f5b354f/tests/framework/db/UnqueryableQueryMock.php#L13-L30

Reply:
It throws the exception in any case. The only way to pass this test is not to call this methods, but we can't achieve it with current realization of emulateExecution